clang/include/clang/Analysis/Scalable/Model/EntityIdTable.h

```diff
@@ -0,0 +1,55 @@
+//===- EntityIdTable.h ------------------------------------------*- C++ -*-===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef LLVM_CLANG_ANALYSIS_SCALABLE_MODEL_ENTITY_ID_TABLE_H
+#define LLVM_CLANG_ANALYSIS_SCALABLE_MODEL_ENTITY_ID_TABLE_H
+
+#include "clang/Analysis/Scalable/Model/EntityId.h"
+#include "clang/Analysis/Scalable/Model/EntityName.h"
+#include <functional>
+#include <set>
+#include <vector>
+
+namespace clang {
+namespace ssaf {
+
+/// Manages entity name interning and provides efficient EntityId handles.
+///
+/// The table maps each unique EntityName maps to exactly one EntityId.
+/// Entities are never removed.
+class EntityIdTable {
```

Comment:
I think this should really be called an `EntityTable` that maps `EntityId` to `Entity` objects.
The `Entity` class should contain `EntityName`, `EntityLinkage`, and other entity properties that will be added by following PRs and future extensions to the framework. Doing this will prevent us from creating separate mappings to associate id with property.

Reply:
Eventually, me might put entities in an equivalence class. In that case, the equivalence class id could be added as a field to the `Entity` object.

Reply:
To close the loop here and to clarify things for others, I will summarize an offline discussion we had on this topic.

We expect this implementation of entity IDs to be used for summary extraction, serialization and deserialization.

For the analysis phase, where we might need to keep in memory all relevant data of all program entities, we expect we will likely develop a different representation of entities. It will need to be reasonably ergonomic for the analysis authors to use, flexible to serve different analyses, and efficient so a whole program analysis of a large program is implementable and performant-enough to be useful. We will possibly take inspiration in clang AST.